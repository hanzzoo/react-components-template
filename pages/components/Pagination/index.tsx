import { useState, useEffect, useCallback, useMemo } from "react";
import { ItemType } from "../../hooks/useFetchData";
import { StyledPagination, StyledPaginationList } from "./style/";

type InitialShowPage = {
  current: number;
  prev: number;
};

export const Pagination = (taskItems: ItemType[]) => {
  const [showPageItems, setShowPageItems] = useState<ItemType[]>([]);
  const initialPageNumbers = [1, 2, 3, 4, 5];
  const prevThreshold = 20;
  const [initialShowPage, setInitialShowPage] = useState<InitialShowPage>({
    current: 0,
    prev: prevThreshold,
  });
  const [currentPage, setCurrentPage] = useState<number>(1);

  useEffect(() => {
    setShowPageItems(
      taskItems.slice(initialShowPage.current, initialShowPage.prev)
    );
  }, [taskItems, initialShowPage]);

  const handleUpdateCurrent = useCallback(
    (event: React.MouseEvent<HTMLElement>) => {
      const dataCurrentAttr = event.currentTarget.dataset.current;
      const currentNumber: number = Number(dataCurrentAttr);
      setCurrentPage(currentNumber);

      if (currentNumber === 1) {
        setInitialShowPage((prevState) => ({
          ...prevState,
          current: 0,
          prev: 20,
        }));
      } else {
        setInitialShowPage((prevState) => ({
          ...prevState,
          current: prevThreshold * currentNumber,
          prev: prevThreshold * currentNumber + prevThreshold,
        }));
      }
    },
    []
  );

  const PaginationController = useMemo(() => {
    return (
      <StyledPagination>
        <li>prev</li>
        {initialPageNumbers.map((num, index) => {
          return (
            <StyledPaginationList
              key={index}
              data-current={num}
              onClick={(event) => handleUpdateCurrent(event)}
              isFocused={currentPage === num}
            >
              {num}
            </StyledPaginationList>
          );
        })}
        <li>next</li>
      </StyledPagination>
    );
  }, [initialShowPage]);

  return {
    showPageItems,
    PaginationController,
  };
};
