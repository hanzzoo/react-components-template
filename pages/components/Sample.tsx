import React from "react";

export const Sample = () => {
  return <h1>welcome to next.js!</h1>;
};
