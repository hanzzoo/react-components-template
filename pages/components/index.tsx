import { useFetchData } from "../hooks/useFetchData";
import { Pagination } from "../components/Pagination";

const Sample = () => {
  const { isLoading, isError, items } = useFetchData();
  const { showPageItems, PaginationController } = Pagination(items);
  console.log(showPageItems);

  return (
    <ul>
      {!isLoading && <p>...Loading</p>}
      {isError && <p>server error</p>}
      {PaginationController()}
    </ul>
  );
};
export default Sample;
