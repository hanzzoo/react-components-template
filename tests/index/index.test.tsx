import { render, screen } from "@testing-library/react";
import { Sample } from "../../pages/components/Sample";

describe("Sample", () => {
  it("renders a heading", () => {
    const { container } = render(<Sample />);

    const heading = screen.getByRole("heading", {
      name: /welcome to next\.js!/i,
    });

    expect(heading).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});
