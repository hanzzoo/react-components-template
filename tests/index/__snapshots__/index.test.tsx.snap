// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders a heading 1`] = `
<div>
  <h1>
    welcome to next.js!
  </h1>
</div>
`;

exports[`Sample renders a heading 1`] = `
<div>
  <h1>
    welcome to next.js!
  </h1>
</div>
`;
